Add JobTaskClusterSpec validate mutator

kanterov opened the pull request:
## Changes
Add JobTaskClusterSpec validate mutator. It catches the case when tasks don't which cluster to use.

For example, we can get this error with minor modifications to `default-python` template:

```yaml
      tasks:
        - task_key: python_file_task
          spark_python_task:
            python_file: ../src/my_project_10/main.py
```

```
 % databricks bundle validate
Error: Missing required cluster or environment settings
  at resources.jobs.my_project_10_job.tasks[0]
  in resources/my_project_10_job.yml:17:11

Task "print_github_stars" requires a cluster or an environment to run.
Specify one of the following fields: job_cluster_key, environment_key, existing_cluster_id, new_cluster.
```

We implicitly rely on "one of" validation, which does not exist. Many bundle fields can't co-exist, for instance, specifying: `JobTask.{existing_cluster_id,job_cluster_key}`, `Library.{whl,pypi}`, `JobTask.{notebook_task,python_wheel_task}`, etc.

## Tests
Unit tests

pietern commented on bundle/config/validate/job_task_cluster_spec.go:
```diff
@@ -0,0 +1,155 @@
+		if task.NotebookTask.WarehouseId != "" {
+			return false
+		} else {
+			return true
```
A notebook task without compute configuration runs on serverless by default.

kanterov replied:
Right. I can add a comment, it still implies that it requires a cluster, right?

pietern replied:
As long as we don't emit a warning for it, yeah.

pietern commented on bundle/config/validate/job_task_cluster_spec.go:
```diff
@@ -0,0 +1,155 @@
+
+			diags = diags.Extend(validateJobTask(rb, task, taskPath))
+		}
+	}
```
This mixes access to the typed configuration with `dyn.Value` access.

Any chance we can just use the `dyn.Value` one?

kanterov replied:
I can change it. I think it's easier to implement and test validations using typed configuration though. The only problematic part is constructing path to a property.

pietern replied:
What I was thinking is you can use `dyn.MapByPattern` to find all tasks across all jobs (and you get the right base path), and then use `convert.ToTyped` to get the typed `jobs.Task` value.

pietern commented on bundle/config/validate/job_task_cluster_spec_test.go:
```diff
@@ -0,0 +1,201 @@
+				TaskKey:         "my_task",
+			},
+			errorPath:    "resources.jobs.job1.tasks[0]",
+			errorDetail:  "Task \"my_task\" has a task type that requires a cluster, but no cluster is specified",
```
What I find tricky (and counterintuitive) is that the error detail talks about a cluster, but the user's underlying intent may be to use serverless and an `environment_key`.

How can we surface this ambiguity in the error/warning? E.g. an and/or?

kanterov replied:
The error summary string contains one of possible attributes to set:
```
Error: One of the following fields must be set: job_cluster_key, environment_key, existing_cluster_id, new_cluster
  at resources.jobs.my_project_10_job.tasks[0]
  in resources/my_project_10_job.yml:17:11

Task "python_file_task" uses a task type that requires a cluster
```

I can change the wording to:
> Task "my_task" has a task type that requires a cluster or environment, but neither is specified.

pietern reviewed (commented):
The code LGTM, but not sure about the classic/serverless ambiguity in the errors.

kanterov commented:
@pietern I've updated error/warning messages and added more comments as you suggested. I assume that the `convert.ToTyped` comment is optional because, at this point, the module is small and logic is encapsulated. I think it makes a lot of sense to have shared visitors, for instance, extract code that visits all tasks that need cluster/environment.

pietern commented:
Yes, it's not strictly necessary.

A higher-level comment about the error messages is that the summary is what users will read first, and that's what should contain what is wrong, not the solution. Right now they are swapped. The solution is in the summary and the error is in the detail. Could you swap those around? Then it will say, `Error: task "main_task" has a task type that requires a cluster or environment, but neither is specified` with the solution following it. WDYT?

pietern commented on bundle/config/validate/job_task_cluster_spec.go:
```diff
@@ -0,0 +1,161 @@
+		} else {
+			// path might be not very helpful, adding user-specified task key clarifies the context
+			detail := fmt.Sprintf(
+				"Task %q requires a cluster or an environment to run. Specify one of the following fields: %s",
```
The line is pretty long like this.

```suggestion
				"Task %q requires a cluster or an environment to run.\nSpecify one of the following fields: %s.",
```

kanterov replied:
I saw it as well.  I think this change works for now. Should we wrap long lines in CLI output? Because we want to reuse the same strings in IDE extensions at some point

pietern reviewed (approved):
Thanks!

Added one more suggestion.

kanterov commented:
@pietern Thanks. I've updated the doc. Can you please add it to the merge queue?
